[BUG] 61.3.0 sdist misses setuptools/tests/config/setupcfg_examples.txt

### setuptools version

61.3.0

### Python version

3.10.4

### OS

Gentoo Linux

### Additional environment information

Using sdist tarball from pypi.

### Description

Many tests fail when running from 61.3.0 due to a missing file, e.g.:

```pytb
________________________________ ERROR collecting setuptools/tests/config/test_apply_pyprojecttoml.py _________________________________
setuptools/tests/config/test_apply_pyprojecttoml.py:21: in <module>
    EXAMPLES = (Path(__file__).parent / "setupcfg_examples.txt").read_text()
/usr/lib/python3.10/pathlib.py:1132: in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
/usr/lib/python3.10/pathlib.py:1117: in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
E   FileNotFoundError: [Errno 2] No such file or directory: '/tmp/setuptools-61.3.0/setuptools/tests/config/setupcfg_examples.txt'
```

### Expected behavior

Tests passing ;-).

### How to Reproduce

```
tar -xf setuptools-61.3.0.tar.gz
cd setuptools-61.3.0
tox -e py310
```

### Output

```console
$ tox -e py310
py310 create: /tmp/setuptools-61.3.0/.tox/py310
py310 develop-inst: /tmp/setuptools-61.3.0
py310 installed: attrs==21.4.0,autocommand==2.2.1,black==22.3.0,build==0.7.0,click==8.1.2,coverage==6.3.2,distlib==0.3.4,docutils==0.18.1,execnet==1.9.0,filelock==3.6.0,flake8==4.0.1,flake8-2020==1.6.1,importlib-metadata==4.11.3,ini2toml==0.10,iniconfig==1.1.1,jaraco.context==4.1.1,jaraco.envs==2.3.0,jaraco.functools==3.5.0,jaraco.path==3.3.1,mccabe==0.6.1,mock==4.0.3,more-itertools==8.12.0,mypy==0.942,mypy-extensions==0.4.3,packaging==21.3,path==16.4.0,pathspec==0.9.0,pep517==0.12.0,pip-run==8.8.0,platformdirs==2.5.1,pluggy==1.0.0,py==1.11.0,pycodestyle==2.8.0,pyflakes==2.4.0,pyparsing==3.0.7,pytest==7.1.1,pytest-black==0.3.12,pytest-checkdocs==2.7.1,pytest-cov==3.0.0,pytest-enabler==1.2.1,pytest-flake8==1.1.1,pytest-forked==1.4.0,pytest-mypy==0.9.1,pytest-perf==0.12.0,pytest-xdist==2.5.0,pytz==2022.1,singledispatch==3.7.0,six==1.16.0,tempora==5.0.1,toml==0.10.2,tomli==2.0.1,tomli_w==1.0.0,tox==3.24.5,typing_extensions==4.1.1,virtualenv==20.14.0,zipp==3.7.0
py310 run-test-pre: PYTHONHASHSEED='525240523'
py310 run-test: commands[0] | pytest
========================================================= test session starts =========================================================
platform linux -- Python 3.10.4, pytest-7.1.1, pluggy-1.0.0
cachedir: .tox/py310/.pytest_cache
rootdir: /tmp/setuptools-61.3.0, configfile: pytest.ini
plugins: xdist-2.5.0, black-0.3.12, cov-3.0.0, flake8-1.1.1, forked-1.4.0, mypy-0.9.1, perf-0.12.0, checkdocs-2.7.1, enabler-1.2.1
gw0 ok / gw1 [1087] / gw2 ok / gw3 ok / gw4 [1087] / gw5 [1087] / gw6 [1087] / gw7 ok / gw8 [1087] / gw9 [1087] / gw10 [1087] / gw11 [1gw0 ok / gw1 [1087] / gw2 ok / gw3 [1087] / gw4 [1087] / gw5 [1087] / gw6 [1087] / gw7 ok / gw8 [1087] / gw9 [1087] / gw10 [1087] / gw1gw0 ok / gw1 [1087] / gw2 ok / gw3 [1087] / gw4 [1087] / gw5 [1087] / gw6 [1087] / gw7 [1087] / gw8 [1087] / gw9 [1087] / gw10 [1087] /gw0 ok / gw1 [1087] / gw2 [1087] / gw3 [1087] / gw4 [1087] / gw5 [1087] / gw6 [1087] / gw7 [1087] / gw8 [1087] / gw9 [1087] / gw10 [108gw0 [1087] / gw1 [1087] / gw2 [1087] / gw3 [1087] / gw4 [1087] / gw5 [1087] / gw6 [1087] / gw7 [1087] / gw8 [1087] / gw9 [1087] / gw10 [1087] / gw11 [1087]
.........................s................F.............s..............................F.F..F...............F..F........F...s.. [ 11%]
.........F.F.F.F...F................F.F.FXF....F...................................FFF....F.....F......................F....... [ 23%]
.............................................................................s....................F.......F...................s [ 35%]
...................................................................................................F........................... [ 46%]
.................F.........................F...F.................................x.F.............x.....x......x......F......... [ 58%]
..........F.....................s.........s............................................................ssss.................... [ 70%]
...............................................x............................................................................... [ 81%]
..................F..............................................F.sss.....................................FF.........F........ [ 93%]
.............................s.sss.sF..F.sFF..s.s.x.x................X.                                                         [100%]/tmp/setuptools-61.3.0/.tox/py310/lib/python3.10/site-packages/coverage/report.py:87: CoverageWarning: Couldn't parse '/tmp/build-via-sdist-5bdmqfw1/ns.pkg-0.0.0/setup.py': No source for code: '/tmp/build-via-sdist-5bdmqfw1/ns.pkg-0.0.0/setup.py'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
[...]

=============================================================== ERRORS ================================================================
________________________________ ERROR collecting setuptools/tests/config/test_apply_pyprojecttoml.py _________________________________
setuptools/tests/config/test_apply_pyprojecttoml.py:21: in <module>
    EXAMPLES = (Path(__file__).parent / "setupcfg_examples.txt").read_text()
/usr/lib/python3.10/pathlib.py:1132: in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
/usr/lib/python3.10/pathlib.py:1117: in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
E   FileNotFoundError: [Errno 2] No such file or directory: '/tmp/setuptools-61.3.0/setuptools/tests/config/setupcfg_examples.txt'
________________________________ ERROR collecting setuptools/tests/config/test_apply_pyprojecttoml.py _________________________________
setuptools/tests/config/test_apply_pyprojecttoml.py:21: in <module>
    EXAMPLES = (Path(__file__).parent / "setupcfg_examples.txt").read_text()
/usr/lib/python3.10/pathlib.py:1132: in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
/usr/lib/python3.10/pathlib.py:1117: in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
E   FileNotFoundError: [Errno 2] No such file or directory: '/tmp/setuptools-61.3.0/setuptools/tests/config/setupcfg_examples.txt'
[...]
```

--- a/MANIFEST.in
+++ b/MANIFEST.in
@@ -15,3 +15,4 @@ include launcher.c
 include msvc-build-launcher.cmd
 include pytest.ini
 include tox.ini
+include setuptools/tests/config/setupcfg_examples.txt

--- a/changelog.d/3233.misc.1.rst
+++ b/changelog.d/3233.misc.1.rst
@@ -1,0 +1,1 @@
+Included missing test file ``setupcfg_examples.txt`` in ``sdist``.

--- a/changelog.d/3233.misc.2.rst
+++ b/changelog.d/3233.misc.2.rst
@@ -1,0 +1,3 @@
+Added script that allows developers to download ``setupcfg_examples.txt`` prior to
+running tests. By caching these files it should be possible to run the test suite
+offline.

--- a/setuptools/tests/config/downloads/.gitignore
+++ b/setuptools/tests/config/downloads/.gitignore
@@ -1,2 +1,4 @@
 *
 !.gitignore
+!__init__.py
+!preload.py

--- a/setuptools/tests/config/downloads/__init__.py
+++ b/setuptools/tests/config/downloads/__init__.py
@@ -1,4 +1,3 @@
-import io
 import re
 from pathlib import Path
 from urllib.request import urlopen

--- a/setuptools/tests/config/downloads/__init__.py
+++ b/setuptools/tests/config/downloads/__init__.py
@@ -9,6 +9,11 @@ NAME_REMOVE = ("http://", "https://", "github.com/", "/raw/")
 DOWNLOAD_DIR = Path(__file__).parent
 
 
+# ----------------------------------------------------------------------
+# Please update ./preload.py accordingly when modifying this file
+# ----------------------------------------------------------------------
+
+
 def output_file(url: str, download_dir: Path = DOWNLOAD_DIR):
     file_name = url.strip()
     for part in NAME_REMOVE:

--- a/setuptools/tests/config/downloads/__init__.py
+++ b/setuptools/tests/config/downloads/__init__.py
@@ -1,0 +1,47 @@
+import io
+import re
+from pathlib import Path
+from urllib.request import urlopen
+
+__all__ = ["DOWNLOAD_DIR", "retrieve_file", "output_file", "urls_from_file"]
+
+
+NAME_REMOVE = ("http://", "https://", "github.com/", "/raw/")
+DOWNLOAD_DIR = Path(__file__).parent
+
+
+def output_file(url: str, download_dir: Path = DOWNLOAD_DIR):
+    file_name = url.strip()
+    for part in NAME_REMOVE:
+        file_name = file_name.replace(part, '').strip().strip('/:').strip()
+    return Path(download_dir, re.sub(r"[^\-_\.\w\d]+", "_", file_name))
+
+
+def retrieve_file(url: str, download_dir: Path = DOWNLOAD_DIR):
+    path = output_file(url, download_dir)
+    if path.exists():
+        print(f"Skipping {url} (already exists: {path})")
+    else:
+        download_dir.mkdir(exist_ok=True, parents=True)
+        print(f"Downloading {url} to {path}")
+        download(url, path)
+    return path
+
+
+def urls_from_file(list_file: Path):
+    """``list_file`` should be a text file where each line corresponds to a URL to
+    download.
+    """
+    print(f"file: {list_file}")
+    content = list_file.read_text(encoding="utf-8")
+    return [url for url in content.splitlines() if not url.startswith("#")]
+
+
+def download(url: str, dest: Path):
+    with urlopen(url) as f:
+        data = f.read()
+
+    with open(dest, "wb") as f:
+        f.write(data)
+
+    assert Path(dest).exists()

--- a/setuptools/tests/config/downloads/preload.py
+++ b/setuptools/tests/config/downloads/preload.py
@@ -1,0 +1,18 @@
+"""This file can be used to preload files needed for testing.
+
+For example you can use::
+
+    cd setuptools/tests/config
+    python -m downloads.preload setupcfg_examples.txt
+
+to make sure the `setup.cfg` examples are downloaded before starting the tests.
+"""
+import sys
+from pathlib import Path
+
+from . import retrieve_file, urls_from_file
+
+
+if __name__ == "__main__":
+    urls = urls_from_file(Path(sys.argv[1]))
+    list(map(retrieve_file, urls))

--- a/setuptools/tests/config/test_apply_pyprojecttoml.py
+++ b/setuptools/tests/config/test_apply_pyprojecttoml.py
@@ -5,7 +5,6 @@ import io
 import re
 import tarfile
 from pathlib import Path
-from urllib.request import urlopen
 from unittest.mock import Mock
 from zipfile import ZipFile
 

--- a/setuptools/tests/config/test_apply_pyprojecttoml.py
+++ b/setuptools/tests/config/test_apply_pyprojecttoml.py
@@ -19,23 +19,23 @@ from setuptools.config import expand
 from setuptools.config._apply_pyprojecttoml import _WouldIgnoreField, _some_attrgetter
 from setuptools.command.egg_info import write_requirements
 
+from .downloads import retrieve_file, urls_from_file
 
+
+HERE = Path(__file__).parent
 EXAMPLES_FILE = "setupcfg_examples.txt"
-EXAMPLES = (Path(__file__).parent / EXAMPLES_FILE).read_text()
-EXAMPLE_URLS = [x for x in EXAMPLES.splitlines() if not x.startswith("#")]
-DOWNLOAD_DIR = Path(__file__).parent / "downloads"
 
 
 def makedist(path, **attrs):
     return Distribution({"src_root": path, **attrs})
 
 
-@pytest.mark.parametrize("url", EXAMPLE_URLS)
+@pytest.mark.parametrize("url", urls_from_file(HERE / EXAMPLES_FILE))
 @pytest.mark.filterwarnings("ignore")
 @pytest.mark.uses_network
 def test_apply_pyproject_equivalent_to_setupcfg(url, monkeypatch, tmp_path):
     monkeypatch.setattr(expand, "read_attr", Mock(return_value="0.0.1"))
-    setupcfg_example = retrieve_file(url, DOWNLOAD_DIR)
+    setupcfg_example = retrieve_file(url)
     pyproject_example = Path(tmp_path, "pyproject.toml")
     toml_config = Translator().translate(setupcfg_example.read_text(), "setup.cfg")
     pyproject_example.write_text(toml_config)
@@ -292,31 +292,6 @@ class TestMeta:
 
 
 # --- Auxiliary Functions ---
-
-
-NAME_REMOVE = ("http://", "https://", "github.com/", "/raw/")
-
-
-def retrieve_file(url, download_dir):
-    file_name = url.strip()
-    for part in NAME_REMOVE:
-        file_name = file_name.replace(part, '').strip().strip('/:').strip()
-    file_name = re.sub(r"[^\-_\.\w\d]+", "_", file_name)
-    path = Path(download_dir, file_name)
-    if not path.exists():
-        download_dir.mkdir(exist_ok=True, parents=True)
-        download(url, path)
-    return path
-
-
-def download(url, dest):
-    with urlopen(url) as f:
-        data = f.read()
-
-    with open(dest, "wb") as f:
-        f.write(data)
-
-    assert Path(dest).exists()
 
 
 def core_metadata(dist) -> str:

--- a/setuptools/tests/config/test_apply_pyprojecttoml.py
+++ b/setuptools/tests/config/test_apply_pyprojecttoml.py
@@ -1,5 +1,7 @@
 """Make sure that applying the configuration from pyproject.toml is equivalent to
 applying a similar configuration from setup.cfg
+
+To run these tests offline, please have a look on ``./downloads/preload.py``
 """
 import io
 import re

--- a/setuptools/tests/config/test_apply_pyprojecttoml.py
+++ b/setuptools/tests/config/test_apply_pyprojecttoml.py
@@ -3,9 +3,11 @@ applying a similar configuration from setup.cfg
 """
 import io
 import re
+import tarfile
 from pathlib import Path
 from urllib.request import urlopen
 from unittest.mock import Mock
+from zipfile import ZipFile
 
 import pytest
 from ini2toml.api import Translator
@@ -18,7 +20,8 @@ from setuptools.config._apply_pyprojecttoml import _WouldIgnoreField, _some_attr
 from setuptools.command.egg_info import write_requirements
 
 
-EXAMPLES = (Path(__file__).parent / "setupcfg_examples.txt").read_text()
+EXAMPLES_FILE = "setupcfg_examples.txt"
+EXAMPLES = (Path(__file__).parent / EXAMPLES_FILE).read_text()
 EXAMPLE_URLS = [x for x in EXAMPLES.splitlines() if not x.startswith("#")]
 DOWNLOAD_DIR = Path(__file__).parent / "downloads"
 
@@ -274,6 +277,18 @@ class TestPresetField:
         reqs = (tmp_path / "requires.txt").read_text(encoding="utf-8")
         assert "importlib-resources" in reqs
         assert "bar" in reqs
+
+
+class TestMeta:
+    def test_example_file_in_sdist(self, setuptools_sdist):
+        """Meta test to ensure tests can run from sdist"""
+        with tarfile.open(setuptools_sdist) as tar:
+            assert any(name.endswith(EXAMPLES_FILE) for name in tar.getnames())
+
+    def test_example_file_not_in_wheel(self, setuptools_wheel):
+        """Meta test to ensure auxiliary test files are not in wheel"""
+        with ZipFile(setuptools_wheel) as zipfile:
+            assert not any(name.endswith(EXAMPLES_FILE) for name in zipfile.namelist())
 
 
 # --- Auxiliary Functions ---

